packages/dds/task-manager/src/taskManager.ts

```diff
@@ -193,6 +211,31 @@ export class TaskManager extends SharedObject<ITaskManagerEvents> implements ITa
             this.removeClientFromQueue(taskId, clientId);
         });
 
+        this.opWatcher.on("complete", (taskId: string, clientId: string, local: boolean, messageId: number) => {
+            if (runtime.connected && local) {
+                const pendingOp = this.latestPendingOps.get(taskId);
+                assert(pendingOp !== undefined, "Unexpected op");
+                // TODO: check below comment and stuff, see if applicable
+                // Need to check the id, since it's possible to complete multiple times before the acks
+                if (messageId === pendingOp.messageId) {
+                    assert(pendingOp.type === "complete", "Unexpected op type");
+                    // Delete the pending, because we no longer have an outstanding op
+                    this.latestPendingOps.delete(taskId);
+                }
+            }
+
+            // We consider locally completed tasks "pending" until they are ack'd. Once they are ack'd the local client
+            // can remove it from this.pendingCompletedTasks. For all other clients, we need to remove the task from
+            // the queue and emit the proper events.
+            if (this.pendingCompletedTasks.has(taskId)) {
+                this.pendingCompletedTasks.delete(taskId);
```

Comment:
I think this "if has, delete" should be in the "if local" check above -- if we see a remote "complete" this should not affect our pending tasks.

We should probably also be tagging the completes with a pendingMessageId like we do in map/directory -- it's maybe not a likely scenario for our target customer, but if a client does a complete-volunteer-complete without waiting for sequencing, then the completes won't get tracked independently and might lead to eventual inconsistency.